### ISSUE TITLE
Annotation tooltip not displayed

### ISSUE THREAD

**Issue by neandersen** [oxyplot/oxyplot#1081]
### Steps to reproduce

I upgraded from oxyplot .core/.wpf 1.0.0-unstable2100 to 2.0.0-unstable0968, no other changes.
My annotation tooltips no longer show, but they did with 1.0.0-unstable2100

1. Create plotmodel and add an annotation to it
2. Set the tooltop for the annotation.
3. Mouse over the annotation

Platform: Windows 10 x64 VS 2017 WPF
.NET version: 4.6

### Expected behaviour

Tooltip shown when mouse hovered over annotation.

### Actual behaviour

Nothing happens

**Comment by silviubogan** [comment]
I see the same bug in the ExampleBrowser, in [this example](https://github.com/oxyplot/oxyplot/blob/develop/Source/Examples/ExampleLibrary/Annotations/AnnotationExamples.cs).

If I create a sample application to test this, I see the same bug, although the property Window.ToolTip works well in the same application, and it is not overriden by another tooltip when the mouse is over the annotation.

I use Windows 10 x64 with VS 2019 and the latest preview version of Oxyplot.Wpf from NuGet at the moment of this writing.

**Comment by silviubogan** [comment]
I wish to solve this issue but I am a beginner. I saw that the ToolTip is actually already set on the WPF elements at the end of the OxyPlot's rendering process so I think it has to do with a detail about how WPF renders elements. If I set the ToolTip property of the Plot element, that ToolTip works. I will analyse this more tomorrow.

**Comment by VisualMelon** [comment]
It seems to be a typical transparent background problem. This can be 'fixed' by removing the background on `mouseGrid` which is presumably eating all the mouse-over events (or setting `HitTestVisible = false`)

The `mouseGrid` was added in #1055 in response to #382 (which notes that the change breaks tooltips): https://github.com/oxyplot/oxyplot/blob/2c254dc8352f59b927e7f6600071fcfa5fe764e7/Source/OxyPlot.Wpf/PlotBase.cs#L292

I'm not a WPF guru, so I don't know a way to resolve the two issues off hand.

**Comment by silviubogan** [comment]
I think about using a new attached property, `Plot.OxyToolTip`, for all the types of annotations and other components of the `Plot` that may have a tooltip to force the display of a `ToolTip` even when the `IsMouseDirectlyOver` property is `false`. For this it may help to use `Reflection` to access the private/protected members of `ToolTip` and related classes, starting [from here](https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/PopupControlService.cs,107). What do you think? Thank you.

**Comment by VisualMelon** [comment]
I'm afraid I've hardly touched WPF for 3 years and don't really know what you mean, but if you can put together a partial solution I'd be happy to look it over. Using reflection to mess with WPF internals doesn't sound ideal. My gut feeling is that there should be a better way to solve #382 so that the existing ToolTip infrastructure can work as intended under WPF; it might even be possible to produce a work-around that changes whether `mouseGrid` (which I think should probably be a member) performs hit-testing or not during plot invalidation. The systems to detect when a plot is updating/has updated are not great at the moment (e.g. #1322), so I might have a look into that.

**Comment by silviubogan** [comment]
The problem is that with, and without, the `mouseGrid`, the annotations and series still do not get mouse events other than MouseDown, so neither MouseMove nor MouseUp. I added [here](https://github.com/oxyplot/oxyplot/blob/f870702/Source/Examples/ExampleLibrary/Series/BarAndColumnSeriesExamples.cs#L372) the following lines:

```c#
s1.MouseDown += R_MouseDown;
s1.MouseUp += R_MouseUp;
s1.MouseMove += R_MouseMove;

s2.MouseDown += R_MouseDown;
s2.MouseUp += R_MouseUp;
s2.MouseMove += R_MouseMove;
```

then [here](https://github.com/oxyplot/oxyplot/blob/f870702/Source/Examples/ExampleLibrary/Series/BarAndColumnSeriesExamples.cs#L945) I added these:

```c#
private static void R_MouseUp(object sender, OxyMouseEventArgs e)
{
    Debug.WriteLine(">>>>> MouseUp");
}
private static void R_MouseMove(object sender, OxyMouseEventArgs e)
{
    Debug.WriteLine(">>>>> MouseMove");
}
private static void R_MouseDown(object sender, OxyMouseDownEventArgs e)
{
    Debug.WriteLine(">>>>> MouseDown");
}
```

The output is like this although I pan/click on the plot/series/annotations:

```
[...]
'ExampleBrowser.exe' (CLR v4.0.30319: ExampleBrowser.exe): Loaded 'C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\System.ServiceModel.Internals\v4.0_4.0.0.0__31bf3856ad364e35\System.ServiceModel.Internals.dll'. Skipped loading symbols. Module is optimized and the debugger option 'Just My Code' is enabled.
>>>>> MouseDown
>>>>> MouseDown
>>>>> MouseDown
>>>>> MouseDown
>>>>> MouseDown
>>>>> MouseDown
>>>>> MouseDown
The program '[0x4DF8] ExampleBrowser.exe' has exited with code 0 (0x0).
```

I wish to test [this solution](https://stackoverflow.com/a/30174766/258462). If this does not work, I think I will debug to see why the mouse events are not being triggered. Besides this, I have code that should work, but I must test to see if the issue #382 does not appear again.

Please tell me with what of what I wrote above you do agree, when you have time. Thank you.

**Comment by VisualMelon** [comment]
It's worth a shot.

I had a go at only enabling the mouse-grid when the plot is invalided, but that didn't seem to help.

**Comment by silviubogan** [comment]
The solution in that SO answer does not work for me.

I've made a scheme because I am new to OxyPlot's source:

![drawing for GitHub 11 09 2019](https://user-images.githubusercontent.com/198924/64670508-0c568a80-d46e-11e9-9cdb-959739c5f639.png)

Nothing registers to the `PlotModel.MouseUp` event. Do you know if this is or is not the problem?

For me, even if I comment all the code inside the method `PlotBase.OnMouseUp`, the series do not receive `MouseUp` or `MouseMove` events.

**Comment by VisualMelon** [comment]
It can be called through the controller, and is exploited in the "Mouse Events/LineSeries and PlotModel MouseDown event" example. The call is indirected through a variable called 'handler' in method `OnMouseUp.OnMouseUp`. _(Side note: all those indirections can be replaced with `?.Invoke` calls, not that this would make it easier to follow the call tree)_

This is the call tree, only the last one is platform specific:

    EventHandler<OxyMouseEventArgs>  UIElement.MouseUp
    virtual void UIElemented.OnMouseUp(OxyMouseEventArgs e)
    virtual void Model.HandleMouseUp(…)
    virtual bool ControllerBase.HandleMouseUp(…)
    void Wpf.PlotBase.OnMouseUp(MouseButtonEventArgs e)

Basically `Model.HandleMouseUp` sends events directly to `UIElements` rather than via the `PlotModel`.

However, I'm not sure what this has to do with tool tips?

**Comment by silviubogan** [comment]
I simply did not know that I have to mark the mouse event handlers with `e.Handled = true` for Series, Annotation etc. so that, for example, after a MouseDown event is triggered, the corresponding MouseUp event's handlers are called too.

I am still not understanding the relationship between Model and PlotModel, but now, that I found the need for `e.Handled = true`, I know what I will retry next.

Below is what I have tried before I reached the conclusion above. There is no need to read this, or answer the question in there.

Thank you.

----

This is my updated scheme:

![drawing for GitHub 11 09 2019 2](https://user-images.githubusercontent.com/198924/64760012-50668f80-d541-11e9-96e6-d2fb731d6332.png)

It shows the problem I am trying to solve. PlotModel.HandleMouseUp calls UIElement.OnMouseUp in the example "LineSeries and PlotModel MouseDown event" but does not call it in my modified "Simple model" example in BarSeriesExamples. It seems to me that in the PlotModel.HandleMouseUp method, when it is called, currentMouseEventElement is null so currentMouseEventElement.OnMouseUp(e) is not called, and this is the reason UIElement.OnMouseUp is not called.

Should I make a separate issue for this?

My changes to the source of BarSeriesExamples' "Simple model" example are [here](https://github.com/oxyplot/oxyplot/issues/1081#issuecomment-529079237).

Relevant lines:

https://github.com/oxyplot/oxyplot/blob/104cfeb56f20cd5309f8f241b2807125fa3d4b52/Source/OxyPlot/Graphics/Model.MouseEvents.cs#L129-L133

**Comment by VisualMelon** [comment]
@silviubogan yes, better to ask there. I was using [this code](https://gist.github.com/VisualMelon/75612b355f43fc46026b0245446bcbc4) (or atleast something similar) to test my attempts, but I don't recall being entirely convinced I was reproducing the same effect. I'll test your code myself tonight if I have time: it would be great to resolve that old issue.

**Comment by silviubogan** [comment]
@VisualMelon 

*Using the latest commit in the `develop` branch and your gist code I get these results:*

The `ToolTip` of the `RectangleAnnotation` is not displayed at all.

*Using my branch and your gist code I get these results:*

The `ToolTip` of the `RectangleAnnotation` is visible for a fraction of second, and then invisible for a fraction of second, and then again visible for a fraction of second and so on.

*In both situations:*

1. The tracker control is not displayed at all.
2. There are a few mouse events lost or received with a delay (I am trying to think what is the difference between lost and delayed).


I think these behaviors are normal because the plot is being invalidated very often. I think these behaviors are acceptable and that with the new changes, OxyPlot is better. Please suggest improvements if you like. If you agree, I will open a pull request.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon 

A screen recording showing my branch and your gist code, a little modified (see the changes below the GIF):

![2019-09-16_09-19-25](https://user-images.githubusercontent.com/198924/64937502-2ecc1780-d863-11e9-83f4-4a66bad27858.gif)

Instead of:

```c#
plot.MouseDown += (s, e) => fs.Title = "" + rnd.Next(100000);
```

I put these:

```c#
plot.MouseDown += (s, e) =>
{
    plot.Title = "MouseDown " + rnd.Next(100000);
    e.Handled = true;
};
plot.MouseUp += (s, e) =>
{
    plot.Title = "MouseUp " + rnd.Next(100000);
    e.Handled = true;
};
plot.MouseMove += (s, e) =>
{
    plot.Title = "MouseMove " + rnd.Next(100000);
    e.Handled = true;
};
```

Thank you.

**Comment by VisualMelon** [comment]
Sorry I didn't get back to you last week.

On my machine (after commenting out `MouseMove` because there are so many events) I'm still losing down and up events. In addition, I don't ever see a tooltip on the rectangle. That might be to do with the relative performance of our computers. Hopefully HermanEldering can help confirm whether this is helping or not, because I don't remember thinking this method is going to work. It may well be that the best we can hope for is to provide a toggle, so that either 'native' tooltips work or you can handling continuous inputs with high-frequency updates. A custom tool-tip system might be in order, as would be needed for #1350.

**Comment by silviubogan** [comment]
@VisualMelon I posted [this comment](https://github.com/oxyplot/oxyplot/issues/382#issuecomment-532066034). I do not want to be impolite, but how many days should we wait for an answer in that issue, #382, before starting to work on one of these from your comment?:

1. a toggle (native tooltips vs. very frequent plot invalidations without tooltips), or
2. a custom tool-tip system.

*Related to (1)*, I wonder whether native tooltips do exclude good results in very frequent plot invalidations. On your computer, does removing the tooltips make fewer mouse events lost?

*Related to (2)*, I think I have a vague idea about how this can be done. I am reading the source  code of OxyPlot, a little bit every day.

I am not sure if we should wait until @HermanEldering answers before I start working on (2), or if I can start working on (2) tomorrow?

Thank you.

**Comment by VisualMelon** [comment]
@silviubogan nothing happens especially quickly on OxyPlot. You've made it as easy as possible, but that is an old issue and it could take time to test the issue properly. In the mean time, you should of course work on whatever is most useful for you.

<hr />

I think it would be worth diverting efforts towards a mechanism which will display tooltips without requiring the support of the render context (i.e. option 2). This would hopefully circumvent the issues with invalidating WPF stuff, and would provide a solution for WinForms and other immediate mode rendering context for which the current mechanism can simply never work (so would be useful whether or not a solution is found for WPF). I think this should be possible without modifying the library code (though it may be desirable to make it part of the library in future). `Popup` looks like a good display solution for a WPF version.

**Comment by silviubogan** [comment]
@VisualMelon 

Looking through the source code I've found these relevant symbols:

1. `PlotElement.ToolTip`
2. `PlotModel.TitleToolTip`
3. `IRenderContext.SetToolTip` (*)
4. `RenderContextBase.SetToolTip` (*)

and in the WPF-specific code:

1. `CanvasRenderContext.ApplyToolTip` (*)
2. `Plot.TitleToolTip`
3. `Series/Annotation/etc.` `.SynchronizeProperties`

From your message I understood that (*) should not be used for `ToolTip`s.

My questions:

1. In WPF [the ToolTip](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.tooltip) can be any type of object, so it is more flexible. Currently the cross-platform ToolTip property is of type string. I would use an object instead and check for string type only for the other platforms, such as WinForms, that need it to be a string. What do you think about this?

2. In the case of WPF, should the value of the native WPF ToolTip property be copied into the internal PlotElement.ToolTip, or we should use a new dependency property such as OxyToolTip from which to copy into the internal ToolTip property?

    - If we use the native ToolTip property we should empty it after copying its contents into PlotElement.ToolTip on [FrameworkElement](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkelement) initialization.
    - If we use OxyToolTip (is there a better name for it?) we should also react to changes to it after initialization, such as 2-way data binding.
    - In both cases, it is necessary to make the native (FrameworkElement.)ToolTip property read-only in the subclasses of FrameworkElement that are part of OxyPlot.Wpf, and I think this is possible by changing their metadata.

The last question (2) has this figure:

![path1056](https://user-images.githubusercontent.com/198924/65372171-1d28ac80-dc75-11e9-932e-c93cc87bb281.png)

Which arrow directions should be allowed/used?

Please let me now if you do not want to tag you in these long messages, or if I should tag someone else in them.

Thank you.

**Comment by VisualMelon** [comment]
Aye, a solution that doesn't require support in the render context is necessary for things like WinForms, and hopefully will provide a workaround for WPF as well. This can use the OxyPlot hit-testing so everything is in model space. This may warrant a review of the update/updating events and all that, which have created issues in the past (though I can't seem to find them right now... #1153 #1322 are relevant). (e.g. if we try to hit-test while the plot is updating we may end up with a jumble; it's not immediately obvious to me how this could work but it can't be worse than what we currently have).

<hr />

1. We should keep it simple and just use strings, certainly to begin with.

2. I'm afraid I don't understand what the arrows mean: are these hypothetical bindings? I have no intuition for WPF/MVVM, so you might have to spell that sort of stuff out for me if you want me to provide any useful comments before it becomes code (which hopefully I can understand)).

    Don't worry about the name for the moment; we can decide that when the thing is working.

<hr />

Feel free to tag me whenever; I read everything on this forum and try to partake whenever I have the energy. Tagging me will only make it easier for me to find stuff I'm meant to be working on in future.

**Comment by silviubogan** [comment]
@VisualMelon I was busy too these days. I am sorry I did not post about it. It may take some time for me to recap.

Yes, the arrows represent data bindings, not implemented with WPF's Binding but through methods like SynchronizeProperties.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon I started working on the WinForms version of OxyPlot [separately](https://github.com/silviubogan/oxyplot/tree/winforms-tooltips). Hit testing and the resulting tooltips work perfectly with annotations, series, the title of the plot view. Now I am trying to do it for the axes and for the legend.

Currently trying to solve this issue: `LinearAxis.HitTest` overrides `UIElement.HitTest` which calls `UIElement.HitTestOverride` which returns `null`. The `LinearAxis` instance is from `this.Model.DefaultXAxis` accessed from `PlotView` from the WinForms version of OxyPlot. It seems to me that `UIElement.HitTestOverride` is always called when doing hit testing, including when using `this.Model.HitTest`.

It seems to me that the legend of the plot is not included in the result of the method `PlotModel.GetHitTestElements`. Should it be included there? The axes are included there.

What is done in my fork's branch [winforms-tooltips](https://github.com/silviubogan/oxyplot/tree/winforms-tooltips), I do not have any idea about how to do it without work inside the library.

Hopefully I did not divert the discussion.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon I have combined all my efforts to make the tooltips work better in [this branch in my fork](https://github.com/silviubogan/oxyplot/tree/crossplatform-tooltip-system).

1. It has the toggle (`UseCustomToolTipSystem`) as a property of `PlotBase`, by default it is `false`.
2. I also added the `Crazy Refresh` example to the Annotations category so I can test with it easily.
3. I see that in the `Crazy Refresh` example the tooltips do not work well neither with `UseCustomToolTipSystem` set to `true`, nor with it set to `false`. My last hope about this is to use the Popup class. But I am still working on (7).
4. I must improve the code style and the similarity between WinForms and WPF versions.
5. The tooltips on the angle/magnitude axes do not work.
6. I might also use the Dispatcher uselessly in some places.
7. There is a visible bug: in the `Annotations > Tool tips` example, when moving the cursor from over the rectangle towards the ellipse, the `ToolTip` is emptied (set to `null`) but a small tooltip appears without contents. If it is not possible to solve this issue w/ `ToolTip` and `ToolTipService`, I will try to use a `Popup`.
8. The hit-testing of axes is not perfect (e.g. when the ticks are inside). This is not very important, I think.
9. There are some constants that may be changed. Here I might need suggestions.

That is it for today.

Thank you.

**Comment by VisualMelon** [comment]
@silviubogan great! I'll take a proper look when I have time. Thanks for working on this.

**Comment by VisualMelon** [comment]
This looks good. I need to take some time to review the WPF and WinForms stuff properly, but this seems to be working well. My ,aom concern is as before, that I don't like the current implementation of `HitTestOverride` and the changes to `GetAxesFromPoint`: an axis should be returning another axis in its `HitTestOverride` method. I'll have a go at re-arranging it myself some time.

Numbered responses:

1. Good. Not sure whether we need something similar for WinForms... probably not.
2. That could be its own PR, if you think it is useful (it is encouraged to provide an example for open issues).
3/4. No comment just yet.
5. I'll take a look.
6. Yeah, looks like it shouldn't be necessary, but I've not tested.
7. Can't reproduce: Is this WPF only?
8. Shouldn't be a blocker.
9. Should probably be configurable somehow. I don't have strong opinions on the defaults.

**Comment by silviubogan** [comment]
@VisualMelon 

>> 7. There is a visible bug: in the `Annotations > Tool tips` example, when moving the cursor from over the rectangle towards the ellipse, the `ToolTip` is emptied (set to `null`) but a small tooltip appears without contents. If it is not possible to solve this issue w/ `ToolTip` and `ToolTipService`, I will try to use a `Popup`.
> 
> Can't reproduce: Is this WPF only?

Yes, it is WPF only.

Thank you.

**Comment by VisualMelon** [comment]
I'll see if I can try it on another machine later today. Might consider adding tooltips to the annotation example if this is all approved.

**Comment by silviubogan** [comment]
@VisualMelon I have solved (7) (the WPF only bug) in [my latest commit](https://github.com/silviubogan/oxyplot/commit/e5606cc7201f093ab816a1de61ff1c3ddbe49aba), but for solving (3) I must try using other members of `ToolTip` and, if this does not work, I will try to use `Popup`.

I do not know why it does work without temporarily enabling `ToolTipService.IsEnabled` when a tooltip should be visible and disabling it after that, but it works.

Tomorrow I will try to find the difference between the purposes of `Plot` and `PlotView` in OxyPlot.Wpf (If you can do this easily, please tell me the difference.) and tackle the other problems I saw in the list in my previous message with a list.

I am glad I can help.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon I solved (3) too in [the latest commit here](https://github.com/silviubogan/oxyplot/tree/crossplatform-tooltip-system). Please continue giving me feedback.

My list of remaining tasks related to these tooltip issues is this:

1. Improve the code style and the similarity between WinForms and WPF versions if and when possible.
2. If possible (and if there is nothing more important than that to do for me for OxyPlot) it would be nice to improve hit-testing of axes to include inner-ticks etc.
3. Make ToolTip constants configurable (`5`, `1000` etc.).
4. When an element is in front of another and it does not have tooltip, the tooltip of the element behind it should be shown... I think this is not done currently, but I am not sure at all.
5. I also want to make a PR with `Crazy Refresh` example, not inside the current branch (I want to remove it from the current branch after making sure the issue is solved and that the example works well).

Thank you for the smooth comunication and for any feedback.

**Comment by silviubogan** [comment]
@VisualMelon 

**Question 1:** I have found this line:

https://github.com/oxyplot/oxyplot/blob/104cfeb56f20cd5309f8f241b2807125fa3d4b52/Source/OxyPlot/Graphics/Model.MouseEvents.cs#L22

Is it correct to make this constant public and to reference it from `PlotBase.HandlePlotElementsToolTip` and `PlotView.HandlePlotElementsToolTip` instead of using the constant `5`? If not, is the value `10` good enough so I can copy-paste it into these methods?

**Question 2:** There are many properties of [the WinForms' ToolTip class](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.tooltip?view=netframework-4.5.2) and of [the WPF's ToolTip class](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.tooltip?view=netframework-4.5.2) / [the WPF's ToolTipService class](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.tooltipservice?view=netframework-4.5.2). Those that have a pair in the other platform are:

WPF | WinForms
-- | --
[ToolTipService.InitialShowDelay](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.tooltipservice.initialshowdelay?view=netframework-4.5.2) | [ToolTip.InitialDelay](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.tooltip.initialdelay?view=netframework-4.5.2)
[ToolTipService.ShowDuration](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.tooltipservice.showduration?view=netframework-4.5.2) | [ToolTip.AutoPopDelay](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.tooltip.autopopdelay?view=netframework-4.5.2)
[ToolTipService.BetweenShowDelay](https://docs.microsoft.com/en-us/dotnet/api/system.windows.controls.tooltipservice.betweenshowdelay?view=netframework-4.5.2) | [ToolTip.ReshowDelay](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.tooltip.reshowdelay?view=netframework-4.5.2)

Also, [the `AutomaticDelay` WinForms property](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.tooltip.automaticdelay?view=netframework-4.5.2) can be simulated using the formulae [here](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.tooltip.automaticdelay?view=netframework-4.5.2) (the same URL as in the previous link) to be usable in OxyPlot.Wpf too.

I think that the most flexible choice is to expose the actual `ToolTip` objects - I've done this in both OxyPlot.Wpf and OxyPlot.WindowsForms. Now we can write code like this on WPF:

```c#
Plot1.Model = ...;
Plot1.OxyToolTip.Background = Brushes.Green;
Plot1.OxyToolTip.Foreground = Brushes.White;
```

or (on WinForms):

```c#
plot1.Model = ...;
plot1.OxyToolTip.IsBalloon = true;
plot1.OxyToolTip.ToolTipIcon = ToolTipIcon.Info;
```

I've done this already, but I think I still have to ask this: is this the best way to go?

**The first new issue:** The tooltips show, in the WPF version, immediately after entering the element with the mouse cursor (*). I must get the default value of the initial delay and use it to make a `DispatcherTimer`. I will be reading for this [this file](https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/ToolTipService.cs,1a3a48f33b29b390) and related files.

**The second new issue:** In WinForms version, sometimes on tooltip opening, for a fraction of second, the previously set tooltip string is visible inside the newly opening tooltip.

Thank you.

### Other details

I do not know if I have put all the matching, good attributes on these 2 properties with the name `OxyToolTip` (e.g. attributes related to WinForms designer and especially the attribute `NonSerialized`, I must see how `NonSerialized` works, also how it works with private fields). Also, I am not sure if the tooltip component in the WinForms version is good to be disposed before the `renderContext` (this is what I've done in the current code) or after.

I construct the WPF `ToolTip` directly in the c-tor of `PlotBase` so code like the one above works well. I see no disadvantage in the place where the `ToolTip` is constructed. I do not know if it would be better to do this or not (or if there is a pattern or recommandation about this): dispose the old ToolTip WinForms object before taking over the new one when using the setter of `OxyToolTip`.

From my last numbered list in my previous message above I tackled (1). I worked on it, but after doing other changes (related to a newly discovered bug in my code (*)) it may be necessary to work on (1) again.

I am not sure if the part of my code using `IsHitTestVisible` property is helping or not. I will figure this out someday.

**Comment by silviubogan** [comment]
@VisualMelon 

Quote from me:

> **The first new issue:** The tooltips show, in the WPF version, immediately after entering the element with the mouse cursor (*). I must get the default value of the initial delay and use it to make a `DispatcherTimer`. I will be reading for this [this file](https://referencesource.microsoft.com/#PresentationFramework/src/Framework/System/Windows/Controls/ToolTipService.cs,1a3a48f33b29b390) and related files.

In today's commit I implemented only the usage of `ToolTipService.InitialShowDelay` in the WPF-based custom tooltip system. What remains to be done is to use in the WPF-based custom tooltip system the attached properties: `ToolTipService.ShowDuration` and `ToolTipService.BetweenShowDelay`.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon For OxyPlot.Wpf, I implemented the usage of all three ToolTipService attached properties (about ToolTip timing) through async-await. It seems to work very well, but I am not 100% sure (I am a very beginner at Task-based programming). I work on the equivalent of this in WinForms, so I can solve this bug:

> **The second new issue:** In WinForms version, sometimes on tooltip opening, for a fraction of second, the previously set tooltip string is visible inside the newly opening tooltip.

I will try not to call Application.DoEvents and to use async-await.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon I made a new commit and I posted [this StackOveflow question](https://stackoverflow.com/q/58267337/258462) in the process of solving this issue:

> **The second new issue:** In [the] WinForms version [of OxyPlot], sometimes on tooltip opening, for a fraction of second, the previously set tooltip string is visible inside the newly opening tooltip.

Thank you.

**Comment by VisualMelon** [comment]
Sorry for the silence: I'm half ill half busy at the moment; I'll be sure to review everything properly over this weekend at the latest.

**Comment by silviubogan** [comment]
@VisualMelon Thank you for letting me know this. I wish you speedy recovery!

**Comment by silviubogan** [comment]
@VisualMelon Today I solved a few issues. A newly discovered unsolved issue is that in the WPF version (and, after a bit of testing, I see it is in the WinForms version too) a rectangle annotation is shown under an axis but the hit-testing (when the mouse is over the axis and the rectangle annotation at the same time) results in the tooltip of the rectangle annotation to be shown or nothing at all.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon The issue in the previous comment is solved.

**Question 1.** Are the tooltips shown only when the mouse position is inside the clip area of the surfaces on which the tooltipped controls are shown? I think yes. Example: bottom axis tooltip == `null`, and a `RectangleAnnotation` is under the bottom axis, totally invisible. The tooltip of the annotation is not shown although the annotation is under the bottom axis which is opaque. I did not investigate this question yet, and I think I do not know from where to start searching. Maybe I'll find the answer tomorrow.

**Question 2.** Should we make an `IToolTippedPlotView` and an `IToolTip` interface common to OxyPlot.Wpf & OxyPlot.WindowsForms?

**Question 3.** Should we handle transparent colors? Example: if one thing is over another opaque thing, the tooltip of the first thing is shown only if its color is not fully transparent; if it is fully transparent, show the tooltip of the opaque thing.

Thank you.

----

This is a scheme I used to correct the hit-testing code [from here](https://github.com/oxyplot/oxyplot/blob/d57391807038bd25c228098249cbc5facd682dda/Source/OxyPlot/PlotModel/PlotModel.cs#L1197-L1234) into [what's here](https://github.com/silviubogan/oxyplot/blob/1d796ea0f562498a9878a13f65ab463c8b5452ce/Source/OxyPlot/PlotModel/PlotModel.cs#L1207-L1250):

![bitmap](https://user-images.githubusercontent.com/198924/66559828-c240f180-eb5e-11e9-91d0-bcea794173f7.png)

**Comment by silviubogan** [comment]
@VisualMelon: **I thought about creating a summary of what you would want to review from my last messages. Please tell me if you want this.** Below is the usual summary for today:

**Question 1** (not related to the ones in the previous messages).

Quote from me:

> If possible (and if there is nothing more important than that to do for me for OxyPlot) it would be nice to improve hit-testing of axes to include inner-ticks etc.

I am not sure if ticks inside the plot area should have the tooltip of the near axis... In the plot area there can be elements, e.g. annotations, that should display their tooltips ignoring the inner-ticks. 
 Hit-testing with each tick without the empty space between them is, I think, bad for performance, but I am not sure. What do you think?

**Question 2.** Another situation: when the `LinearAxis.TitlePosition` is `0` or `1`, or other close values, the tooltip is not always displayed over the axis title label. Should I look more into this? Please let me know if you have advice for me. I would consider more doing the hit-testing correctly for this case, but **(a)** the `ClipTitle` property of the `LinearAxis` seems to not work (neither in WPF nor in WindowsForms, including in the first screenshot below) and **(b)** there are situations like this (and I am not sure if it is the best to cut through the corners of the control with a diagonal starting from the axes intersection, or if this slows down the hit-testing too much):

![2019-10-11_12-11-07](https://user-images.githubusercontent.com/198924/66644135-dbae7000-ec28-11e9-9b8a-4e063cd7e074.png)

**Question 3.** In the Universal Windows solution I am an absolute beginner and, probably, I do not even have installed all the necessary components, so this might be simple. The `SimpleDemo` project builds successfully with a very little change: in *PlotView.cs* add this:

```c#
public bool UseCustomToolTipSystem
{
    get
    {
       return true;
    }
}
```

but I cannot run it, neither with debugging enabled, nor with debugging disabled... Neither with the `Simulator` nor with the `Local Machine`. When starting the app, it says in the status bar of VS 2019 that the build was successful: "Build succeeded", and this is the final message:

![2019-10-11_11-29-50](https://user-images.githubusercontent.com/198924/66644178-f7197b00-ec28-11e9-87bd-a7839613b6c4.png)

In the Output pane I get these lines:

```
1> OxyPlot.Windows -> G:\Lucru\GitHub\silviubogan-oxyplot\Source\OxyPlot.Windows\bin\Debug\OxyPlot.Windows.dll
2>------ Skipped Build: Project: SimpleDemo, Configuration: Debug x86 ------
2>Project not selected to build for this solution configuration
3>------ Skipped Deploy: Project: SimpleDemo, Configuration: Debug x86 ------
3>Project not selected to build for this solution configuration
========== Build: 1 succeeded, 0 failed, 1 up-to-date, 1 skipped ==========
========== Deploy: 0 succeeded, 0 failed, 1 skipped ==========
```

**Question 4.** I see that hit-testing code for `RectangleAnnotation` returns the annotation even if the mouse cursor is over the axes or even beyond them (see [here](https://github.com/oxyplot/oxyplot/blob/d4aaaf5ef0039f87a2f8dd0fd94da96cf7b10481/Source/OxyPlot/Annotations/RectangleAnnotation.cs#L134-L149)). I see, after a fast scan, that the same situation is in `EllipseAnnotation`, `PathAnnotation` and `TextualAnnotation` too.

**Example [here](https://gist.github.com/silviubogan/ab01f35a12761fabd4b6696837b5d091):** make 4 `LinearAxis`s, one for each side of the plot; make a big rectangle annotation that is beyond the upper-right corner of the plot area but clipped. See that the tooltip of the rectangle can be visible when putting the cursor over the zone of axes intersection that is not took into account when hit-testing for axes.

![2019-10-11_12-57-04](https://user-images.githubusercontent.com/198924/66644379-78710d80-ec29-11e9-91bf-200fd62e34fe.png)

I have not looked yet in other `PlotElement` subclasses, but I still ask: should I modify, for each plot element that needs hit-testing, this method (`HitTestOverride`) or make another hit-testing method made especially for tooltips, based on the `HitTestOverride`?

Thank you.

**Comment by VisualMelon** [comment]
This seems to be working well.

Comments on comments above:

 - I'm afraid I also have no idea how to use UWP, but it would be more than useful to get WPF and WinForms working.

 - For axes tooltips, I would keep it simple and 'assume' the axis is a rectangle defined by its start position, end position, and 'depth'. I don't think hit-testing inner-ticks is a good idea, because it adds complexity and means the axis extends into the plot area (in the non-zero-crossing case, which probably needs testing at some point). I think the current behaviour is good.

 - With regards to hit-test ordering, it's probably best to mirror `GetHitTestElements()` in `PlotModel`.

 - The hit-test constants should be configurable (discussed below)

Comments on code:

 - There is a lot of overlap in the WPF and WinForms code, and it is non-trivial code. It would be if these could be pulled out into a class which managed all the delays and cancellation, so that the many members involved do not pollute the `PlotView`s and the logic can be changed in one place only. This class can house the constants and such. An `IToolTipView` or something that can be implemented per-view and provides stuff like `Show()`, `Hide()`, `Text` would seem reasonable.

 - `ToolTippedPlotElement` (if it is necessary) should be packaged away with the aforementioned class (both could be `Utilities`.
    - Members of course need documentation at some point
    - I'd be tempted to remove the cross-type comparisons, because they aren't being used and add stuff that needs to be maintained. I'd prefer well-named method for such operations, which are less surprising. I don't like that `Equals` has different behaviour: that's even more surprising.
    - `GetHashCode` should be implemented: it can basically just return the hash code of the appropriate member.

 - I'm not sure when `UpdateToolTip();` should be called, but I think the current behaviour is acceptable (e.g. not sure if panning/zooming should update the tooltip).

 - I don't like that `UseCustomToolTipSystem` is in `IPlotView`: if it needed, then it should be a concrete property of the individual plot view implementations. If we find a need to move it into `IPlotView` later, then that can be done.

I haven't properly reviewed the code in the `PlotView` which does the actual work. I might have a go at extracting a `ToolTipManager` class or something myself, so that I'm forced to understand it fully. With all your work on hit-testing within the main library, I can't immediately see why there should need to be any changes to the PlotViews themselves: the tooltip management can be completely separate and just hook the relevant events. If that can be done, then it should be trivial to reuse it in the future, and means that there is an extensible mechanism available for people to modify if they need to adapt it without having to touch library code.

A comment about git: you should really have rebased your code onto the resent addition to develop, as this will be necessary when it comes to the pull-request. You should be able to sort this out by just resetting to your last commit and rebasing from there.

Thanks again for your hard work!

**Comment by silviubogan** [comment]
@VisualMelon I see that the tooltips in both WinForms and WPF work well when panning/zooming: the tooltip almost always reflects the `OxyToolTip` string of the `PlotElement` below the mouse cursor, with the exception being the situation presented in [the StackOverflow question linked above](https://stackoverflow.com/q/58267337/258462).

I started working on the `IToolTip` interface and implementations. A WinForms prototype is in the latest commit in my fork's branch `crossplatform-tooltip-system`. I put the `IToolTip` in the `Utilities` directory. I also must put XML comments in the interface as well, I think. The implementations of `IToolTip`, which I think I will call both `CustomToolTip`, will both have a public property `NativeToolTip` so that the WinForms `ToolTip` object or the WPF `ToolTip` object will be accessible. I also put a `Dispose` method in the interface, because in WinForms it must be disposed (I am not sure if I should implement `IDisposable` or not). In the WPF `CustomToolTip` the `Dispose` method will be empty, I think.

I plan to expose the customizable tooltip properties in the `IToolTip` interface. Currently the `IToolTip.Text`'s property is not working as a simple property: it hides/shows the tooltip not only when the new value is different than the previously set string: it hides when the value is null, and shows when the previously hovered plot element is different than the currently hovered plot element, regardless of the `Text` property value. Maybe there are also other problems too.

I did not tackle yet the issues regarding `ToolTippedPlotElement` and I started using the git rebase command.

Thank you for the feedback and for the kind words. I am glad I can help.

**Comment by silviubogan** [comment]
@VisualMelon I've made both `CustomToolTip` classes that implement `IToolTip`. The current WPF implementation of `IToolTip` (`CustomToolTip`) subscribes to `MouseEnter`/`MouseLeave`/`MouseMove` events of the `PlotBase`. The new issue is with the test posted [here](https://gist.github.com/silviubogan/ab01f35a12761fabd4b6696837b5d091) and in the WPF-based `CustomToolTip`. Maybe I can make the event somewhere to be marked not handled, so that these events reach the handlers in `CustomToolTip`? I think that the only other solution is a new event, and I do not know how it should be called.

* *With the current solution*, registering to these events in the c-tor, I get this behavior: the tooltip does not update while dragging the cursor from a plot element with a tooltip string `X` to another plot element with a tooltip string `Y` where `X != Y`.

* *With this code in PlotBase.Events.cs*:

```C#
/// <summary>
/// Invoked when an unhandled MouseMove attached event reaches an element in its route that is derived from this class. Implement this method to add class handling for this event.
/// </summary>
/// <param name="e">The <see cref="T:System.Windows.Input.MouseEventArgs" /> that contains the event data.</param>
protected override void OnMouseMove(MouseEventArgs e)
{
    base.OnMouseMove(e);
    (OxyToolTip as CustomToolTip).Pb_MouseMove(this, e);
    if (e.Handled)
    {
        return;
    }

    e.Handled = this.ActualController.HandleMouseMove(this, e.ToMouseEventArgs(this));
}
```

instead of

```C#
/// <summary>
/// Invoked when an unhandled MouseMove attached event reaches an element in its route that is derived from this class. Implement this method to add class handling for this event.
/// </summary>
/// <param name="e">The <see cref="T:System.Windows.Input.MouseEventArgs" /> that contains the event data.</param>
protected override void OnMouseMove(MouseEventArgs e)
{
    base.OnMouseMove(e);
    if (e.Handled)
    {
        return;
    }

    e.Handled = this.ActualController.HandleMouseMove(this, e.ToMouseEventArgs(this));
}
```

and with these lines commented out from the `CustomToolTip` c-tor:

```
this.pb.MouseMove += Pb_MouseMove;
this.pb.MouseLeave += Pb_MouseLeave;
this.pb.MouseEnter += Pb_MouseEnter;
```

and with the `Pb_MouseMove` handler marked as public, the bug is not taking place (but the code is very unmantainable).

I think I will investigate this more tomorrow. My to-do list includes making `ToolTippedPlotElement` a little better, as you suggested, and exposing through IToolTip the 3 common `ToolTip` properties in WPF and WinForms, and the hit-testing tolerance variable.

Thank you. I wish you good health.

**Comment by silviubogan** [comment]
@VisualMelon

Today's subissue is this: in both the WinForms code and in the WPF code, although I use an intermediary `CancelableTaskAsync` method that should handle these errors, I still get many messages exactly like this one in the Output pane:

> Exception thrown: 'System.Threading.Tasks.TaskCanceledException' in System.Private.CoreLib.dll

I think that maybe it is just a beginner problem and that I will investigate this more tomorrow.

My list of remaining tasks related to the tooltip issues (**the things not restated here are solved in my opinion**):

- [ ] 1. Small issue:

```c#
<param name="el">The <see cref="PlotElement"/>.</param>
```

`PlotElement` is detected for Visual Studio info tooltip as the property, not the type. (in ToolTippedPlotElement.cs, when hovering with the mouse over `PlotElement` in that context).

- [ ] 2. Order alphabetically the members of the new interface and of the new classes.
- [ ] 3. Improve XML comments in other files than IToolTip.cs and the two CustomToolTip.cs.
- [ ] 4. A little reminder if I can find something useful on Google: try to implement hit-testing with polar plots when it makes sense.
- [ ] 5. Expose in IToolTip the 3 common tooltip properties that are available both in WPF and in WinForms under some different names.
- [ ] 6. Should IToolTip inherit from IDisposable?
- [ ] 7. I am not sure if the part of the new code in OxyPlot.Wpf using the WPF property `IsHitTestVisible` (the `mouseGrid` related stuff) is helping or not, so I might remove my changes there.
- [ ] 8. Remove the changes in my branch in these files:

* [Source/Examples/ExampleLibrary/Annotations/AnnotationExamples.cs](https://github.com/silviubogan/oxyplot/blob/b921ded7e494743282a04f93c984f20627ed559e/Source/Examples/ExampleLibrary/Annotations/AnnotationExamples.cs)

* [Source/Examples/WindowsForms/WindowsFormsDemo/Form1.cs](https://github.com/silviubogan/oxyplot/blob/b921ded7e494743282a04f93c984f20627ed559e/Source/Examples/WindowsForms/WindowsFormsDemo/Form1.cs)

- [ ] 9. Maybe I used the Dispatcher uselessly in some places in the WPF version.

I hope this helps.

Thank you.

**Comment by silviubogan** [comment]
@VisualMelon I have marked as checked some issues in the check list in my previous message (I will try to keep the checkboxes updated). And a new issue is that while the tooltip is open it does not get updated when changing the tooltip of the hovered `PlotElement` (*at least in the WPF version, I did not test the WinForms version yet*).

To make this work I would need an event for when the tooltip string is changed (if there is no `PropertyChanged` general event like in `INotifyPropertyChanged`). The `PlotElement` class has a `ToolTip` auto-property but does not have a `PropertyChanged` event or something similar. **What would be a walkaround to this problem?** I also think that `PlotModel.TitleToolTip` should be watched for changes too.

I see that this can be done in client code *for WPF*, but it still has some small bugs visible to me (1. leaving the window with the mouse cursor lets the tooltip there, 2. sometimes moving the cursor over the tooltipped series does not show the tooltip, 3. the tooltip flashes with the same contents even if the string is not changed):

![2019-10-17_12-57-21](https://user-images.githubusercontent.com/198924/66999154-2a9f4e00-f0de-11e9-9aad-fce53974701f.gif)

For this I still had to change some things *in the WPF* `CustomToolTip.cs` and in `IToolTip.cs`. For it to work I need to commit a few new changes (still experimenting) such as the `IToolTip.Show()` and `IToolTip.Hide()` to return `bool`, not `void`: whether the respective operation was completed successfully. I think I will commit these changes when the WinForms version is tested too.

Please let me know (when you have time) if I post too often (I really do not want to spam).

Thank you.

**Comment by VisualMelon** [comment]
I don't think watching the tool-tip text is a good idea; if anything, it would make sense to listen for plot update events, and re-do the hit-testing and reapply the tool-tip text if necessary (a simple comparison should avoid any flickering). Tooltips are not really designed to be used on crazily dynamic things: watching for updates should be technically easy to implement, and then easy to disable if people don't want it.

I don't mind the updates at all - especially as you are soliciting feedback in 'substantial' posts - but I can't speak for the 168 other watchers.

I'll take a proper look at the code again soon, and hopefully we can get some other feedback soon.

**Comment by VisualMelon** [comment]
Looking at your latest commits, if there isn't enough redundancy in `CustomToolTip` to warrant a separate controller, then I'm not sure the PlotViews need to be aware of it, and the `IToolTip` interface would seem somewhat redundant. It would be better, I think, to separate the view and the controller (would mean using the Model events rather than View, if possible), but I don't think that should be necessary to get this included.

The `PlotView`s probably shouldn't take an `IToolTip` in the constructor (and `IToolTip tt` could do with a better name); I'm not sure how based to resolve this, but for the moment I would be inclined to require manually attaching the ToolTips (i.e. the PlotView never creates one for you).

<hr />

@objorke Would you be happy to take a look at this some time soon? At be909138, the changes are basically:
 - `OxyPlot.Core`
   - Modify `PlotModel` to accommodate Hit-testing for axes
   - Add Hit-testing for axes (doesn't work with polar plots at the moment)
   - Add some stuff to `Utilities` to support the implementation of 'Custom' ToolTips on a per-platform basis
 - `OxyPlot.WindowsForms` and `OxyPlot.WPF`
   - Implement `CustomToolTip`, which hooks relevant events in `PlotView` and manages a tooltip accordingly.

The changes to `OxyPlot.Core` can be made independently of the changes to WPF/Winforms, and would - in my opinion - be valuable in their own right.

All the non-trivial logic is in `CustomToolTip`, and I'll review this properly again once @silviubogan is happy with his work.

**Comment by silviubogan** [comment]
@VisualMelon I do not know if I should sort the class members somehow. I do not understand how a Controller could be extracted from the CustomToolTip (if you wish, I can study [this]( https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) more). In rest, I am happy with my work, if you do not have more feedback to give.

* [Here](https://github.com/silviubogan/oxyplot/tree/crossplatform-tooltip-system) is the branch with the changes I made, and
* [here](https://github.com/silviubogan/oxyplot/tree/testing-crossplatform-tooltip-system) is another branch, based on the first, with WPF and WinForms examples and demos, just for testing.

Thank you.

**Comment by VisualMelon** [comment]
@silviubogan 
Concerning ordering, the best thing to do would be to add the StyleCop analyser (from Sam Harwell (sharwell)) and see if it tells you to rearrange anything. The old StyleCop system had a particular order ordering for the sort of member (e.g. field, static field, constructor, etc.), but the order within these sorts would seem even more subjective.

The idea of the controller would be that the View would be specific to each PlotView, and would provide a simple interface like "ShowToolTip(string position, ScreenPoint location); HideToolTip();" and the controller would handle all the events and such from the PlotModel, perform the ToolTip logic (e.g. deciding to hide it when the mouse moves, deciding to show it after 2 seconds or whatever) and issue instructions to the View. That way, all the complex code you have written to handle the ToolTip can be kept in one place (the common Controller class) and be used with any PlotView that has a suitable ToolTipView. This means the cost of supporting each platform will be lower, as bug fixes to the Controller won't have to be duplicated, and only logic specific to the platform needs to be written for each platform. It also means the Controller can be swapped out by consumer without having to produce per-platform TooltipViews, which reduces there costs also, and means they don't have to keep a custom version of `WpfToolTipView` up-to-date, they only need to focus on their `CustomToolTipController`.

I'll try to take a sensible length of time to look it all over soon, but don't want to promise anything before the weekend.

**Comment by silviubogan** [comment]
@VisualMelon Today's progress (**no question, nothing urgent**) on my side is this: I have started working on the Controller in [this branch](https://github.com/silviubogan/oxyplot/tree/tooltips-controller). I will try to make it work with both WinForms and WPF, with the same code. After finishing the work, including the work on the WPF version, I will use StyleCop again to remake the code style. I also had to make an addition to `IPlotView`, to get the `ScreenPoint` of the mouse cursor. I will look later to find if it is possible to get the mouse position with already existing code.

So now I have two interfaces: `IToolTipView` and `IToolTipController`, and two classes that inherit from one, respective, another. One is, for example, `WinFormsToolTipView` (and in future `WpfToolTipView`), another is `ToolTipController`. The two interfaces and the `ToolTipController` are put in the `Utilities` directory inside the cross-platform library.

The `PlotView`'s ctor does not accept a tooltip view (not even a controller) as a parameter. It creates one when its `Model` is set.

Thank you for all the given feedback and for the figure (it is very nice in my humble opinion).

**Comment by silviubogan** [comment]
@VisualMelon In the WPF version, I create the `ToolTipController` and its `WpfToolTipView` only in `PlotView`, not in `PlotBase`, because in `PlotBase` I did not find something similar to `OnModelChanged()` (I think that the `Controller` should be recreated every time the plot model reference is changed).

The related remaining problem is that changing the tooltips view, the tooltips controller or the plot model, in some combination, after the plot view has been initialized, does not work, I think. I think that I will investigate this more if you do not disagree.

Thank you.

----

Other small details: after the tooltip is shown over a `PlotElement`, if I wait a few seconds for it to disappear then move the mouse over the same `PlotElement`, the tooltip (with the previous string) does not show up again. No tooltip shows up again. I think this should remain as expected behavior because I tested this with the mouse cursor over the Start button in Visual Studio, and its tooltip has the same behavior.

**Comment by VisualMelon** [comment]
The `Controller` work looks good. Here are some comments/questions:
 - The `ToolTippedPlotElement(bool)` constructor is a bit dodgy, since the parameter _has_ to be `true` otherwise the element is left as null. It might make sense to provide a single `protected` constructor, and provide 2 well-named static methods for the tasks of creating one with a `PlotElement` and one without.
 - Since you've got the `IToolTipController` interface, it would be worth using it for the `ToolTipController` properties of the Winforms and WPF PlotViews so that it is swappable.
 - I'm not sure I see the need for `GetClientScreenPointForMouse`: I think you can just pass the `ScreenPoint` to `UpdateToolTip` from the `OxyMouseEventArgs`. Am I missing something? It would be great if we didn't have to modify `IPlotView`, and minimised the amount of supporting code that needs to be maintained per-PlotView.
 - Could do with a comprehensive example in the example library. This will make it much easier to test implementations for different platforms.

There is one big thing left, I think: there needs to be a way to attach and detach the tooltip 
controllers. This probably has to be done in the PlotView. It's important that they can be swapped in and out without interfering, and that it's possible to have none at all for performance purposes (hit-testing is expensive, and if none of my series have tooltips then I'd rather not have the overhead). I'm off the opinion now that 'on by default' is fine if it is easy to disable.

Once that's sorted, I think this should be pretty much ready for a PR, though I'm sure I've said that before ;)

<hr />

@objorke would you want this broken down into separate PRs? The way I see it there are 4 parts to this (assuming we don't have to modify `IPlotView`), which could be their own PRs, but I don't know if that is worth the effort (is it useful somehow?):
 - Changes to the Core library to improve hit-testing support
 - Additions to the Core library of the `ToolTipController` and related interfaces
 - Additions and to WPF and Winforms

**Comment by objorke** [comment]
Impressing work, @silviubogan! Sorry, I have not been able to follow all discussion here.
I think this should be a single PR, it is all related to the same issue, right? We could also create a "preview PR" to make it easier to see all the changes.

**Comment by silviubogan** [comment]
@objorke I want to thank you for the message. It means a lot to me.

I am very sorry, I should have posted a more recent message with another link to another branch on which you could have made the pull request. I do not know if it is possible for you to change the branch in the preview pull request. If it is, please change it to [`crossplatform-tooltip-system`](https://github.com/silviubogan/oxyplot/tree/crossplatform-tooltip-system). If not, I can merge [`crossplatform-tooltip-system`](https://github.com/silviubogan/oxyplot/tree/crossplatform-tooltip-system) into the branch you used, `tooltips-controller`. I am waiting for your advice on this.

Thank you again.

@VisualMelon I think that tomorrow I will analyze your message and I will start to work on the tasks that you gave me in your message. Thank you.

**Comment by objorke** [comment]
I created a new PR on the correct branch :-)

Sorry I have created some extra work for you by moving WPF classes to oxyplot-contrib...
Or maybe this is better, you don't need to handle changes to WPF XAML classes in this PR

**Comment by silviubogan** [comment]
@VisualMelon Please forgive me, today I managed to do just a new rebase and a forced push. Thank you for all the advice.

**Comment by trovialdo** [comment]
I did not follow the discussion (nor reviewed the PR) above and just saw the issue.

I have locally a different (basic) implementation for ToolTips, not sure it covers everything but it works well for me so far, the gist:
- Remove tooltips from IRenderContext
- On IPlotView add: `void ShowToolTip(string text, ScreenPoint position);`
- On Model.MouseEvents add:  `protected abstract void ShowToolTip(object sender, OxyMouseEventArgs e);`
- In Model.MouseEvents change:
```
protected virtual void OnMouseMove(object sender, OxyMouseEventArgs e)
        {
            try
            {
                this.MouseMove?.Invoke(sender, e);

                this.ShowToolTip(sender, e); // TODO should this be called from the controller?
            }
            catch (Exception exception)
            {
                e.Handled = true;
            }
        }

```
```
        protected virtual void OnMouseLeave(object sender, OxyMouseEventArgs e)
        {
            try
            {
                this.MouseLeave?.Invoke(sender, e);

                this.ShowToolTip(sender, e); // To make sure the tool-tip is hidden
            }
            catch (Exception exception)
            {
                e.Handled = true;
            }
        }
```
- In PlotModel add:
`public bool AreToolTipsEnabled {get; set; }`
```
protected override void ShowToolTip(object sender, OxyMouseEventArgs e)
        {
            if (this.AreToolTipsEnabled) // TODO this should be determined by the controller
            {
                try
                {
                    var screenPosition = e.Position;
                    string tooltipText = null;

                    // To reset the tool-tip when the mouse leaves the plot (required for some niche cases where otherwise the tool-tip persists)
                    if (!new OxyRect(0, 0, this.Width, this.Height).Contains(screenPosition))
                    {
                        PlotView.ShowToolTip(null, screenPosition);
                        return;
                    }

                    if (this.TitleArea.Contains(screenPosition))
                    {
                        tooltipText = this.TitleToolTip is null ? (this.TitleToolTipFormatter is null ? null : this.TitleToolTipFormatter(this)) : this.TitleToolTip;
                    }
                    else
                    {
                        // TODO this can be optimized if we do not hit test elements without tooltip/-formatter set
                        var args = new HitTestArguments(screenPosition, MouseHitTolerance);
                        foreach (var result in this.HitTest(args))
                        {
                            tooltipText = result.Element.ToolTip is null ? (result.Element.ToolTipFormatter is null ? null : result.Element.ToolTipFormatter(result)) : result.Element.ToolTip;
                            if (tooltipText != null)
                            {
                                break;
                            }
                        }
                    }
                    this.PlotView.ShowToolTip(tooltipText, screenPosition);
                }
                catch (Exception exception)
                {
                    // TODO show to user
                }
            }
        }
```
- On PlotElement add:  `public Func<HitTestResult, string> ToolTipFormatter { get; set; }` so the user can generate the tooltip on the fly if desired (see logic in the function above)
- For WPF in PlotBase add:
```
        public void ShowToolTip(string text, ScreenPoint position)
        {
            if (string.IsNullOrEmpty(text))
            {
                if (ToolTip is ToolTip t)
                {
                    t.IsOpen = false;
                }
                ToolTip = null;
            }
            else
            {
                var toolTip = ToolTip as ToolTip;
                if (toolTip != null)
                {
                    if ((string)toolTip.Content == text && toolTip.IsOpen)
                    {
                        return;
                    }

                    toolTip.IsOpen = false;
                }

                ToolTip = new ToolTip()
                {
                    Content = text,
                    IsOpen = true,
                    Placement = System.Windows.Controls.Primitives.PlacementMode.MousePoint, // Should be the same as the position
                    VerticalOffset = 4,
                    HorizontalOffset = 4
                };
            }
        }
```

I was not planning to make a PR yet and will now leave it to silviubogan. Just wanted to bring it up here since it might give some inspiration. I particularly would like to push to add the ToolTipFormatter functionality.

**Comment by silviubogan** [comment]
@trovialdo Thank you for the contribution. If @VisualMelon agrees with the `ToolTipFormatter` I can implement it this in the existing PR. From what I understand, it is implemented in the `PlotElement` class and each instance of the `PlotElement` has its own tooltip formatter. Thank you again.

**Comment by trovialdo** [comment]
That is correct, when the ToolTip property is not set the ToolTipFormatter is used instead (if set). By default it is null.

**Comment by silviubogan** [comment]
@VisualMelon

> Since you've got the `IToolTipController` interface, it would be worth using it for the `ToolTipController` properties of the Winforms and WPF PlotViews so that it is swappable.

I think this is already done.

Questions:

- [ ] Is `ToolTipFormatter` from @trovialdo a good idea in your opinion and should I work on it?
- [ ] Should I also make an example for when the tooltip contains the same random number generated in the last mouse event, not a newly generated number like in the current example?

My check list:

- [ ] The last big thing you mentioned in your message (a way to attach and detach the tooltip
controllers).
- [ ] I still have to make changes to some files in OxyPlot.Wpf.Contrib.
- [ ] Check if code-gen works well in the 3 new examples.
- [ ] Rerun StyleCop.
- [ ] If there is time for this and you agree, I wish to test again if the tooltip show/hide timing is well configured (I used a few async-await calls). I think that it probably is, and this is not hard to test.

Thank you.

**Comment by VisualMelon** [comment]
@trovialdo thanks for chiming in!

<hr />

@silviubogan
> Is ToolTipFormatter from @trovialdo a good idea in your opinion and should I work on it?

I'd have no objection to adding a formatter to the default tool tip controller. It shouldn't be a priority, because - if I'm understanding correctly - it is basically only a concern for the Controller, so should be easy to add later and people can use a custom controller in the interim if they wish. Equally, it shouldn't be too complex, so if you are happy to work on it, I'd say go for it. If possible, a format-string style would be nice as well (because this is provided throughout the library already), but again that should be easy to add later.

> Should I also make an example for when the tooltip contains the same random number generated in the last mouse event, not a newly generated number like in the current example?

If you think that would be useful, go for it. It need not be separate example, just a separate annotation in an existing one, if you are worried about creating lots of boring and repetitive code.

> I think this is already done.

Sorry, I might have been looking at the wrong branch.

<hr />

Now that we have https://github.com/oxyplot/oxyplot/pull/1415/files/3c14084ae565f17b1a5c65a958c70fb0e0c25c94..8aa3f4c1cac57f4b725f8c2ea2e2d0058fd0aa9b I would suggest we move discussion about the implementation over there: it'll be much easier to review the code there. (obviously I'll still keep an eye on anything here, but unless its related to this issue was it was intended, I'll be posting over there)

**Comment by objorke** [comment]
I added some comments in the "preview" PR. Sorry I have not had time to read all in this thread yet, so I may have misunderstood some things. My primary concerns are
- don't depend on the events on the `PlotModel`, I still like the idea of #111
- don't add new interfaces where new functionality could naturally belong to existing interfaces (`IView` for the show/hide tooltip)
- can we remove the tooltip functionality from the `RenderContext`?
- it is not really needed to create a new property for the tooltipcontroller, it could be integrated with the plot controller - but I am not sure which pattern would be the best
I really like that the control of tooltips are now in a separate class and can be shared across the platforms!

**Comment by VisualMelon** [comment]
> can we remove the tooltip functionality from the RenderContext?

I think it should remain: exports could include the tooltip, but naturally cannot provide an interactive experience to go along with it.

<hr />

Regarding the other points, would the idea then be to integrate `IToolTipView` into `IView`, and hand control of the `IToolTipController` off to the `IController` (rather than the PlotView/nobody in particular)? I hope I haven't mislead Silviubogen by trying to encourage as much decoupling from the main library as possible; if this has your full support as an integral part of the library, then perhaps a breaking change would be alright.

**Comment by objorke** [comment]
@trovialdo I just discovered your comment and suggestions - I like your simple solution!

I have two comments
- I don't think we need to add an `AreToolTipsEnabled` property - can't we simply not specify any tooltips?
- move the `ShowToolTips` event handler from the Model.MouseEvents to the `ControllerBase` class, and rename it to avoid confusion with a `ShowToolTip` method of `IView`

**Comment by objorke** [comment]
@VisualMelon I understand the idea about supporting tooltips in exports, I understand this can be implemented in SVG and PDF exporters
- https://developer.mozilla.org/en-US/docs/Web/SVG/Element/title
- https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf
I agree it could be a good idea to keep this possibility. We could add two issues to implement tooltips on SVG and PDF exports.

But how to avoid conflict with the "tooltip-controller" solution?

**Comment by objorke** [comment]
This issue is about a simple bug on WPF, but I think this thread has moved on to new features

- support tooltips on Windows Forms (and make it easy to support tooltips on other platforms)
- move implementation of tooltips from rendering code to view/controller
- support for dynamic tooltips (this is getting close to the functionality of "trackers", should we restrict tooltips to be static?)

Can we create new issues for these changes/new features, and keep this issue for the simple "annotations not displayed" bug? The annotations were working in the past, without need for architecture changes.

**Comment by objorke** [comment]
@VisualMelon cohesion and loose coupling is good, but I think it can also increase the complexity of the solution. I would like to keep the code as simple as possible for everybody to understand. We should careful when adding or changing public interfaces and base classes to reduce the risk for introducing breaking changes later.

I think that a tooltip belongs to the view and nowhere else, and it should be only 1 tooltip visible at the same time - so this could simply belong to the `IView` interface.

The code controlling the tooltips is also something we just need one implementation and one instance of, and we don't really need public interfaces for this. If not "customizability" is required, I think we should avoid adding offering that option. I think the tooltip control code can be implemented internally on the `ControllerBase` without any changes to public interfaces.

**Comment by VisualMelon** [comment]
@objorke my original suggestion was to make the work that Silviubogan is now doing a 'bolt on', so if you couldn't use the `RenderContext` based tooltips you had an alternative. I don't see why both can't be provided, and just leave the choice of which one gets used to the application/PlotView.

**Comment by VisualMelon** [comment]
@objorke Whatever you think is best; I just suggested the looser version because I wanted to avoid the risk that it couldn't be included in the main library for some reason, and to avoid modifying any existing interfaces. It all depends on how you interpret `IView` really, and if you think requiring tooltip provisioning is a good idea then I'm more than happy to go along with it.

**Comment by objorke** [comment]
I created some new issues
- the cross-platform feature: #1418 
- hit testing on axes: #1419 
- hit testing on titles: #1420

I hope this can make it easier to get these changes merged! :-)

**Comment by trovialdo** [comment]
@objorke Thanks! It is only a quick and easy solution though, to get them working, i would not advise to actually use said implementation as is since it will completely flood (temporary freeze) the UI with invisible (discarded) tooltips if you generate a tooltip it in the `Delta` of a `MouseManipulator` (bound in the `PlotController`) (although that is specific to the WPF implementation in the `IView` and not the architecture). I never put any thought into it how to resolve it since they were already working on another, more fleshed out, solution for the tooltips in OxyPlot.
* The `AreToolTipsEnabled` was added to be able to disable tooltips if not used (it requires a hittest every time regardless whether or not one wants a tooltip -> performance).
* I agree, in that implementation, `ShowToolTips` should be moved to the 'ControllerBase'

**Comment by mennowo** [comment]
Are there plans to actually implement the changes proposed in this thread? Would be nice to have tooltips for annotions.
